ui/summary.go

```diff
@@ -31,99 +30,119 @@ import (
 
 	"github.com/loadimpact/k6/lib"
 	"github.com/loadimpact/k6/stats"
+	"github.com/pkg/errors"
 	"golang.org/x/text/unicode/norm"
 )
 
 const (
-	GroupPrefix   = "█"
-	DetailsPrefix = "↳"
+	groupPrefix   = "█"
+	detailsPrefix = "↳"
 
-	SuccMark = "✓"
-	FailMark = "✗"
+	succMark = "✓"
+	failMark = "✗"
 )
 
 var (
-	ErrStatEmptyString            = errors.New("invalid stat, empty string")
-	ErrStatUnknownFormat          = errors.New("invalid stat, unknown format")
-	ErrPercentileStatInvalidValue = errors.New("invalid percentile stat value, accepts a number")
+	errStatEmptyString            = errors.New("invalid stat, empty string")
+	errStatUnknownFormat          = errors.New("invalid stat, unknown format")
+	errPercentileStatInvalidValue = errors.New("invalid percentile stat value, accepts a number")
 )
 
-var TrendColumns = []TrendColumn{
-	{"avg", func(s *stats.TrendSink) float64 { return s.Avg }},
-	{"min", func(s *stats.TrendSink) float64 { return s.Min }},
-	{"med", func(s *stats.TrendSink) float64 { return s.Med }},
-	{"max", func(s *stats.TrendSink) float64 { return s.Max }},
-	{"p(90)", func(s *stats.TrendSink) float64 { return s.P(0.90) }},
-	{"p(95)", func(s *stats.TrendSink) float64 { return s.P(0.95) }},
+// ErrInvalidStat represents an invalid trend column stat
+type ErrInvalidStat struct {
+	name string
+	err  error
 }
 
-type TrendColumn struct {
-	Key string
-	Get func(s *stats.TrendSink) float64
+func (e ErrInvalidStat) Error() string {
+	return errors.Wrapf(e.err, "stat '%s'", e.name).Error()
 }
 
-// VerifyTrendColumnStat checks if stat is a valid trend column
-func VerifyTrendColumnStat(stat string) error {
-	if stat == "" {
-		return ErrStatEmptyString
+// Summary handles test summary output
+type Summary struct {
+	trendColumns        []string
+	trendValueResolvers map[string]func(s *stats.TrendSink) interface{}
+}
+
+// NewSummary returns a new Summary struct. It returns an error if the
+// passed stats fields cannot be used as a trend column.
+func NewSummary(sts []string) (*Summary, error) {
+	tc := []string{"avg", "min", "med", "max", "p(90)", "p(95)"}
+	if len(sts) > 0 {
+		tc = sts
 	}
 
-	for _, col := range TrendColumns {
-		if col.Key == stat {
-			return nil
-		}
+	resolvers := map[string]func(s *stats.TrendSink) interface{}{
+		"avg":   func(s *stats.TrendSink) interface{} { return s.Avg },
+		"min":   func(s *stats.TrendSink) interface{} { return s.Min },
+		"med":   func(s *stats.TrendSink) interface{} { return s.Med },
+		"max":   func(s *stats.TrendSink) interface{} { return s.Max },
+		"p(90)": func(s *stats.TrendSink) interface{} { return s.P(0.90) },
```

Comment:
Are these actually necessary, given the fact that we have a way to generate arbitrary percentiles in `generateCustomTrendValueResolvers()`?

Reply:
I considered that, but decided against it since `p(90)` and `p(95)` will pretty much always be the default percentiles, so it doesn't make sense to conditionally create them. This way it clearly shows what the defaults are.

Reply:
Yeah I am wondering the same ... given the code below, won't removing those two line .... continue to work as before ?

Reply:
@MStoykov Yes, it will, but like I said, I think it's clearer defining them upfront. Generating them muddles up the default vs. custom concept (though we could just call it `generatePercentileResolvers`), and it would have to go through `validatePercentile` for no reason.

But let me know if you feel strongly about it and I'll remove them.

Reply:
To be honest I think that https://github.com/loadimpact/k6/pull/1143/files#diff-cc7288c8b0cb64e706ac188a9e524b7fR82 should be set in https://github.com/loadimpact/k6/pull/1143/files#diff-b5b20214939c9d689266a9dccbd0c47cR66 as the default and it should this should not be named `getDefaultResolvers`, but something like: getCustomResolvers/ getNonPercentileResolvers ?

Reply:
@MStoykov we can't set the default in the CLI flags, because then we'd have the exact same issue we had with the `--system-tags` @cuonglm fixed in https://github.com/loadimpact/k6/pull/1092

Reply:
>.> ah .. :disappointed:  well .. then I still think that this should be renamed to ... `getNonPercentileResolvers`, and possibly moving those defaults to ... somewhere else ;) and showing them as the defaults in the cli. Maybe we should just do what was done in @1092 for those as well ..

Reply:
OK, I'm a bit lost :smile: Can you clarify on how to proceed here?

1. Remove the percentile columns and let them be auto-generated with the user-specified ones? I'm fine with pre-defining them as it's clearer and they won't have to run through `validatePercentile` in the vast majority of use cases, but let me know if you still want them removed.
2. Define default trend columns as default arguments to `--summary-trend-stats`, and apply the #1092 fix. I think this makes sense, as the stats vs. columns distinction is currently a bit muddy, and this change would make it clearer that these are stat names that map to metric resolver functions (whose values are then rendered as columns).
3. Rename `getDefaultResolvers` to something else. I don't see a reason for this, as these are, well, the default resolvers and not custom (those are the percentiles generated by what the user specified), and not sure about having "Non" in the name. How about `getTrendValueResolvers`? Not sure, you suggest something, I'm fine with `getDefaultResolvers` :D.

Let me know if this covers everything, and your thoughts/decision.

Reply:
I would think that with #961 we are probably going to add some way for all the non percentile ones to be parsed directly from the name like `avg` or even `avg+count`(totally useless). (and yes I know that the issue is about thresholds parsing not summary columns but I would argue that code will be reused). 
At which point this map will be ... removed :). So from that perspective I don't know if it's actually worth the effort.
At least for me (@na-- might still think that it's worth to do most of the mentioned things above now), adding the defaults to the message is probably the only thing that we should do right now. Everything else seems (given that issue I linked and it's apparent relevance to this code) not too important ... especially given that it creates even more code to be refactored when we start to fix the configuration ... 

On the other hand I do agree with the first comment of @na-- as this lines are definitely not needed and again ... the defaults are defined few lines after this ... we just aren't calling them that, there ;)

Reply:
Yeah, to me `getDefaultResolvers()` doesn't make a lot of sense. We basically have 2 things:
1. The possible values that can be configured as trend columns in the summary. This includes a mix of static (`min`, `max`, etc.) and dynamic ( `p(??)`) columns, as well as default and non-default (`count`) values so each value should be validated by a function. And each value corresponds to a "resolver", i.e. something that returns its actual value.
2. The default values that will be displayed - this is an **ordered** subset of 1, so an unordered map can't be it (and why `getDefaultResolvers()`) is a bad name. This should be just a simple list (i.e. a slice), and yeah, the best UX would be to display its default value it in the CLI help, so the https://github.com/loadimpact/k6/pull/1092 workaround may be necessary :disappointed:

Reply:
OK, that sounds good.

So to summarize what needs to be done in this PR:

1. Remove the pre-generated percentile resolvers and rename `getDefaultResolvers` to ...? @na-- you need to be the tie-breaker here with a good name suggestion, since @MStoykov doesn't like mine and I don't like his, so... :smile:

2. Move the definition of columns to show as the default value for `--summary-trend-stats`, and apply the #1092 fix (whatever that is, still need to grok how it works).

Would that cover it?

Reply:
1. Remove the pre-generated percentile resolvers and rename `getDefaultResolvers` to `getStaticResolvers()`, or to just heave it a simple `map`, since this map will always be the same... 
2. The #1092 fix is that:
   - we set the CLI flag default value to `nil`, but we manually set the help text there to include the actual default value (see `systemTagsCliHelpText` in that PR)
   - we have an `applyDefault` function that is called at the end of the config consolidation - it applies the default value for any complex types (`SystemTags` in that PR, `SummaryTrendStats` in this) that are still nil (i.e. that the user hasn't specified)

Reply:
Let me know if b91229ab resolves this.

There are some new globals (I kept in line with the `SystemTags` implementation) and ad-hoc repeated validation and just general suckiness, but the configuration system is a pain to work with...

And I need to rebase and fix the merge conflicts, but I'm waiting on the final approvals before doing so.